### PR TITLE
🚨 Alerting 

### LINKED ISSUE #36: [OPE-38] Notification System
Whenever a ping is **not** `200`\-`299`, we should send a notification to the user.

Before starting, we should make sure to have a stable API around that.

For each integration, we should have a `mapping` function from our API to the service one.

<sub>From [SyncLinear.com](https://synclinear.com) | [OPE-38](https://linear.app/openstatus/issue/OPE-38/notification-system)</sub>

### LINKED ISSUE #38: [OPE-41] Integrate Email
<sub>From [SyncLinear.com](https://synclinear.com) | [OPE-41](https://linear.app/openstatus/issue/OPE-41/integrate-email)</sub>

### PR DESCRIPTION
**Pull request by thibaultleouay**
Still a WIP 
But a workspace can set up multiple notification 
We should create something in the settings for that 

Mail would be enable by default for every workspace 

We should allow the user to as many notification as he wants ( we will have more provider after )

We should call the send method from the providers 

When you set up a  monitors you can select many notification you have set up for your workspace

### CODE REVIEW

File: apps/web/src/app/api/checker/regions/_checker.ts [hunk 1]
```diff
@@ -121,3 +123,23 @@ export const ping = async (
...
+    )
+    .leftJoin(
+      schema.monitor,
+      eq(schema.monitor.id, schema.notificationsToMonitors.monitorId),
+    )
+    .where(eq(schema.monitor.id, Number(monitorId)))
+    .all();
+
+  for (const notif of notifications) {
+    // Send the appropriate notification
```
Comment (thibaultleouay) [inline apps/web/src/app/api/checker/regions/_checker.ts]: @mxkaske Here i want to call the the send method from the  appropriate package 
e.g  notification.name = 'email' -> call openstatus/resend/send.ts 
notification.name='discord' -> call openstatus/discord/send.ts

File: packages/emails/index.ts [hunk 2]
```diff
@@ -1,12 +1,19 @@
 import type { ReactElement } from "react";
 import { Resend } from "resend";
 
+import Alert, { EmailDataSchema } from "./emails/alert";
```
Comment (mxkaske) [inline packages/emails/index.ts]: Let's call it `AlertEmailDataSchema` as we will have multiple `EmailDataSchema`s in the future.

File: packages/db/drizzle/0008_square_madelyne_pryor.sql [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+	FOREIGN KEY (`workspace_id`) REFERENCES `workspace`(`id`) ON UPDATE no action ON DELETE no action
+);
+--> statement-breakpoint
+CREATE TABLE `notifications_to_monitors` (
+	`monitor_id` integer NOT NULL,
+	`notificationId` integer NOT NULL,
+	PRIMARY KEY(`monitor_id`, `notificationId`),
+	FOREIGN KEY (`monitor_id`) REFERENCES `monitor`(`id`) ON UPDATE no action ON DELETE no action,
+	FOREIGN KEY (`notificationId`) REFERENCES `notification`(`id`) ON UPDATE no action ON DELETE no action
+);
```
Comment (mxkaske) [inline packages/db/drizzle/0008_square_madelyne_pryor.sql]: Reducing complexity by removing different notifications for monitors. The defined notification channels by the user are used to send informations when a monitor is down.

File: packages/db/src/schema/notification.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+import { monitor } from "./monitor";
+import { workspace } from "./workspace";
+
+const name = ["email", "discord", "slack"] as const;
+
+export const notification = sqliteTable("notification", {
+  id: integer("id").primaryKey(),
+  name: text("name", { enum: name }).notNull(),
+  data: text("data", { mode: "json" }).notNull().default("{}"),
```
Comment (mxkaske) [inline packages/db/src/schema/notification.ts]: Is `data` where we will store information for the different notifications? Like _email address_ for "email", _webhook url_ for "slack",...

**Comment by thibaultleouay** [comment]
if we detect an error we set the monitor status to error and send an email 
if we reping and the status is now running we update the status

File: apps/web/src/app/api/checker/utits.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import type { z } from "zod";
+
+import type { notificationName } from "@openstatus/db/src/schema";
+import { send as sendEmail } from "@openstatus/resend";
+
+// FIXME: Better type
+export const mapTypeToFunction: {
```
Comment (thibaultleouay) [inline apps/web/src/app/api/checker/utits.ts]: @mxkaske  if you find a smart way to do that 🔥

File: packages/db/src/schema/notification.ts (hunk 4 above)
Comment (thibaultleouay) [reply packages/db/src/schema/notification.ts]: yes

**Pull request merged**